Launch startup board for first time users #1243

codemanmeet opened the pull request:
Fixes #1243. Also updated sample board filters.

damithc commented on src/main/java/ui/TestController.java:
```diff
@@ -65,6 +66,12 @@ public static boolean isTestGlobalConfig() {
         return hasUI() && commandLineArgs.getOrDefault("testconfig", "false").equalsIgnoreCase("true");
     }
 
+    // used for testing the creation of startup board on startup (a.k.a when the number of repos is 0 initially)
```
Write proper javadoc comments.

codemanmeet replied:
Sure!

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -143,6 +143,16 @@ private void login(boolean isBypassLogin) {
         getMainWindowHandle(mainStage.getTitle());
     }
 
+    private void startupBoardInitiator(boolean shouldCreateSampleBoard){
```
Instead of passing this boolean, simply not call this method if it is false?

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -133,7 +131,7 @@ private void createWorkAllocationBoard() {
     }
 
-    private void createSampleBoard() {
+    public void createSampleBoard(boolean isCalledOnStartup) {
```
Needs a javadoc comment. Also, the parameter can be named after what it does instead of how it's used. suppressDialog?

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -150,7 +148,10 @@ private void createSampleBoard() {
         triggerBoardSaveEventSequence(SAMPLE_BOARD);
 
-        DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD, SAMPLE_BOARD_DIALOG);
+        //isCalledOnStartup is set to true when called on launch in UI.java
```
We shouldn't make reference to how methods are used in comments.

dariusf commented on src/main/java/ui/TestController.java:
```diff
@@ -65,6 +66,14 @@ public static boolean isTestGlobalConfig() {
+    /** Used for testing startup board on launch (i.e. when the number of stored repos is 0)
+     * StartBoardLauncherTest makes use of this argument to test the launch of the board for first time users
+     */
+    public static boolean isTestStartupBoard() {
```
Again, we shouldn't refer to how methods are used. The description should start on the second line, as a matter of style.

Also, the method name isn't very clear, even though I understand what it does from the method body. `shouldTestStartupBoard`?

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -143,6 +143,14 @@ private void login(boolean isBypassLogin) {
+    private void launchSampleBoard(){
+        BoardAutoCreator boardCreator = new BoardAutoCreator(this, panels, prefs);
+        if (prefs.getAllBoardNames().contains(boardCreator.SAMPLE_BOARD)){
+            prefs.removeBoard(boardCreator.SAMPLE_BOARD);
```
Maybe you could explain why this part is needed with an example in a comment? Is there a case where it could remove something legitimate?

codemanmeet replied:
I was thinking maybe if we wanted to update the sample board in the future the user could get a new version of the sample board. But now that I think, the user himself would then not be a first time user and he certainly would not delete repo files from the store folder himself. The `Auto-create` itself would enable the creation of a new sample board if need be! I guess this if statement is not necessary. I'll get rid of it.

codemanmeet replied:
I was also thinking of possibility of duplicates, which is also not possible since this is a first time user we are talking about!

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -150,6 +158,7 @@ private void disableUI(boolean disable) {
     }
 
     private void showMainWindow(String repoId) {
+        boolean shouldCreateSampleBoard = logic.getStoredRepos().size() == 0;
```
`logic.getStoredRepos().isEmpty()`

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +183,9 @@ private void showMainWindow(String repoId) {
         // Should only be called after panels have been initialized
         ensureSelectedPanelHasFocus();
         initialisePickers();
+        if (shouldCreateSampleBoard && (!TestController.isTestMode() || TestController.isTestStartupBoard())){
```
Can the other parts of the conditional be part of `shouldCreateSampleBoard`?

It might be helpful to explain why we don't want to open the sample board when in test mode, but without the flag.

Also, we should move that variable down here, since it's only used here.

codemanmeet replied:
Startup board testing is required only on app launch. Other tests go with the usual assumption of bypassing login and testing HubTurbo features assuming a repo has been created. Thus, we need a flag to test the startup board feature. Makes it neater. Yup, it's helpful to explain that. I have added a comment about it too.

`logic.openPrimaryRepository`/`logic.setDefaultRepo` change `logic.getStoredRepos().isEmpty()` so this needs to be tested before those two. The number of stored repos will be 0 the first time a user logs in (i.e.) if the store folder is empty. So the check is needed at that place.

dariusf commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,62 @@
+        type("test");
+        click("Sign in");
+
+        PlatformEx.waitOnFxThread();
```
Why is this needed?

dariusf commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,62 @@
+        PlatformEx.waitOnFxThread();
+
+        click("File");
+        click("Logout");
```
There's a UITest method for traversing menus which would help here.

codemanmeet replied:
Will make use of `clickMenu`.

dariusf commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,62 @@
+        click("File");
+        click("Logout");
+
+        assertEquals(panelControl.getPanelCount(), SAMPLE_PANEL_NAMES.size());
```
Maybe we should check the panel names as well? Since they're static. Just to ensure that they end up being used in the board that's open.

codemanmeet replied:
Will do that!

codemanmeet commented:
@dariusf Have addressed the comments! Let me know if it can be merged.

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -150,7 +152,9 @@ private void createSampleBoard() {
         triggerBoardSaveEventSequence(SAMPLE_BOARD);
 
-        DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD, SAMPLE_BOARD_DIALOG);
+        if (!suppressDialog){
```
As the boolean is inverted here, why not use `isShowDialog` in the first place?

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -133,7 +131,11 @@ private void createWorkAllocationBoard() {
+     * Creates a sample board to showcase HubTurbo's functionality with sample repos in filters.
+     * @param suppressDialog
+     */
+    public void createSampleBoard(boolean suppressDialog) {
```
This is not how we name booleans :-|

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +180,12 @@ private void showMainWindow(String repoId) {
+        //Startup board testing required only on app launch.
+        //Checks if the number of repos is 0 indicative of first time users of the app.
+        if (shouldCreateSampleBoardOnAppLaunch && (!TestController.isTestMode()
+                || TestController.shouldTestStartupBoard())){
```
The line wrapping should be at the highest level possible, in this case, at `&&`. The current line wrapping has less readability.

dariusf commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -150,7 +152,9 @@ private void createSampleBoard() {
         triggerBoardSaveEventSequence(SAMPLE_BOARD);
 
-        DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD, SAMPLE_BOARD_DIALOG);
+        if (isShowDialog){
```
Maybe this could be `isDialogShown`?

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +180,12 @@ private void showMainWindow(String repoId) {
+        //Startup board testing required only on app launch.
+        //Checks if the number of repos is 0 indicative of first time users of the app.
+        if (shouldCreateSampleBoardOnAppLaunch &&
+                (!TestController.isTestMode() || TestController.shouldTestStartupBoard())){
```
Could this be factored out into a variable and named? Would be a bit easier to read.

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +181,12 @@ private void showMainWindow(String repoId) {
+        //Startup board testing required only on app launch.
+        boolean shouldBeMainLaunchOrStartupBoardTesting =
+                !TestController.isTestMode() || TestController.shouldTestStartupBoard();
+        if (shouldCreateSampleBoardOnAppLaunch && shouldBeMainLaunchOrStartupBoardTesting){
```
I feel like these names are at too low a level of abstraction to help readability. The first comment is unnecessary if we use it as the variable name.

```java
boolean isFirstTimeUser = logic.getStoredRepos().isEmpty();
```

I get that giving a succinct name to the `not test mode || has flag` condition is difficult, so we can just give it a general name, and explain why we need it with a header comment (all tests start with a clean slate and we don't want the sample board then, etc.).

```java
if (isFirstTimeUser && shouldOpenSampleBoard()) {
    // ...
}

boolean shouldOpenSampleBoard() {
    return !TestController.isTestMode() || TestController.shouldTestStartupBoard();
}
```

Exact names can be improved, but the level of abstraction should be similar.

dariusf commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +181,12 @@ private void showMainWindow(String repoId) {
+        boolean shouldBeMainLaunchOrStartupBoardTesting =
+                !TestController.isTestMode() || TestController.shouldTestStartupBoard();
+        if (shouldCreateSampleBoardOnAppLaunch && shouldBeMainLaunchOrStartupBoardTesting){
+            launchSampleBoard();
```
openSampleBoard maybe? More consistent with the terms we use in the UI.

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -150,7 +152,9 @@ private void createSampleBoard() {
-        DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD, SAMPLE_BOARD_DIALOG);
+        if (isDialogShown){
+            DialogMessage.showInformationDialog("Auto-create Board - " + SAMPLE_BOARD, SAMPLE_BOARD_DIALOG);
+        }
```
This behavior is not documented in the header comment.

codemanmeet replied:
Added it

damithc replied:
Why do we need a dialog here? New users are going to see this dialog?

codemanmeet replied:
The dialog is needed when the sample board is created using `Auto Create` menu. The dialog is not seen by new users.

damithc replied:
Got it. Showing the dialog is an additional step that you do at the end, so an alternative implementation could be like this:
```
void createSampleBoard(){...}

void createSampleBoardAndInformUser(){
  createSampleBoard();
  //show dialog
}
```
Using boolens to control behaviors is not wrong, but not pretty either.
It's ok to keep the current one.

codemanmeet replied:
Okay, got it.

damithc commented on src/main/java/ui/TestController.java:
```diff
@@ -65,6 +66,14 @@ public static boolean isTestGlobalConfig() {
 
+    /**
+     * Tests the launch of the sample board on first time app startup.
+     */
```
The comment doesn't match the method. The method doesn't test anything.

codemanmeet replied:
changed.

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -143,13 +143,29 @@ private void login(boolean isBypassLogin) {
+     */
+    private boolean shouldOpenSampleBoard() {
+        return !TestController.isTestMode() || TestController.shouldTestStartupBoard();
+    }
```
Any way to avoid/minimize polluting production code with test-specific code like this?

codemanmeet replied:
Yup moved it out to `TestController.java`

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -143,13 +143,29 @@ private void login(boolean isBypassLogin) {
+    }
+
     private void showMainWindow(String repoId) {
+        boolean isFirstTimeUser = logic.getStoredRepos().isEmpty();
```
Declare variable as close to where it is needed as possible.

codemanmeet replied:
This is the closest to where it can be. This has to be here since `logic.getStoredRepos()` gets altered in the next line by `logic.openPrimaryRepository`.

codemanmeet commented:
@dariusf @damithc Made all the relevant refactoring changes! Can I go ahead and merge now?

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -27,30 +27,28 @@
-            String.format("repo:%s;%s is:unread updated<24 sort:!updated",
-                    FIRST_SAMPLE_REPO_NAME, SECOND_SAMPLE_REPO_NAME)
+            String.format("repo:%s is:open is:issue !label:priority.", FIRST_SAMPLE_REPO_NAME)
             ));
```
Instead of two parallel `List` objects, isn't it better to use one `Map` object?

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -133,7 +131,12 @@ private void createWorkAllocationBoard() {
-    private void createSampleBoard() {
+    /**
+     * Creates a sample board to showcase HubTurbo's functionality with sample repos in filters.
+     * The confirmation dialog pops up for the creation of the sample board if isDialogShown is set to true.
```
What's the purpose of the dialog box?

codemanmeet replied:
The dialog box confirms the creation of the sample board and tells the user it's been saved.

codemanmeet commented:
@damithc I have changed the two lists for sample panel names and sample panel filters to a hashmap and updated header comments. Let me know if the PR is good now.

dariusf commented on src/main/java/ui/TestController.java:
```diff
@@ -86,6 +95,15 @@ public static boolean isCloseOnQuit() {
+     * so don't need boards created on startup.
+     */
+    public static boolean shouldOpenSampleBoard() {
+        return !TestController.isTestMode() || TestController.shouldTestStartupBoard();
```
We no longer need to reference TestController here.

codemanmeet replied:
Will remove the reference!

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -143,6 +143,11 @@ private void login(boolean isBypassLogin) {
         getMainWindowHandle(mainStage.getTitle());
     }
 
+    private void openSampleBoard(){
```
`open` or `create`?

codemanmeet replied:
The sample board is opened as well as saved in the config file (when it is created and opened). Let's go ahead with `openAndSaveSampleBoard`.

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -150,6 +155,8 @@ private void disableUI(boolean disable) {
     }
 
     private void showMainWindow(String repoId) {
+        boolean isFirstTimeUser = logic.getStoredRepos().isEmpty();
```
Move this near to where the variable is used (i.e. line 184).

codemanmeet replied:
As mentioned before in the review, this has to be on this line as `logic.getStoredRepos()` is altered by `logic.openPrimaryRepository` which is just after this line. We want to capture the boolean value right on this line.

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -174,6 +181,9 @@ private void showMainWindow(String repoId) {
         initialisePickers();
+        if (isFirstTimeUser && TestController.shouldOpenSampleBoard()){
+            openSampleBoard();
+        }
```
May be this problem goes beyond this PR, but the dependency from this class to `TestController` should be in the other direction.

codemanmeet replied:
Yup, this problem transcends beyond this PR. TestController is used throughout `UI.java`. Would have to create a separate issue for this. Will go ahead and do that.

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -105,14 +104,21 @@ public void sampleBoardAutoCreationTest() {
+        int j = 0;
+        for (String panelFilter : BoardAutoCreator.getSamplePanelDetails().values()) {
+            assertEquals(panelInfos.get(j).getPanelFilter(), panelFilter);
+            j++;
```
What is being tested here? Can you add a comment to explain?

codemanmeet replied:
This was a refactoring done to convert those two `Lists` to one `Map`. Will add a comment for testing here.

damithc replied:
I don't think you need two loops to iterate a map.
On a related note, I think we need classes to represent information of a board (`BoardInfo`?), and a panel (`PanelInfo`?) The current code is getting messy due to lack of right abstractions.

codemanmeet replied:
We already have `panelInfo` class created. I'm making use of it as well. `BoardInfo` would include a list of panelInfos. This would result in quite a bit refactoring for the existing code on boards, which is beyond the scope of this PR. 

I am now using one loop to iterate through the map.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,71 @@
+        for (String panelFilter : BoardAutoCreator.getSamplePanelDetails().values()) {
+            assertEquals(panelInfos.get(j).getPanelFilter(), panelFilter);
+            j++;
+        }
```
This code block seems like a copy-paste. Can reuse instead?

codemanmeet replied:
Yup, we can reuse :+1:

damithc commented on src/main/java/ui/BoardAutoCreator.java:
```diff
@@ -133,15 +126,20 @@ private void createWorkAllocationBoard() {
-    private void createSampleBoard() {
+    /**
+     * Creates a sample board to showcase HubTurbo's functionality with sample repos in filters.
+     * The dialog box shows a confirmation message for the saving of the sample board if isDialogShown is set to true.
```
The header comment says `saving of the sample board` but the dialog is not about saving the sample board right?

codemanmeet replied:
yup, the dialog is about creating and loading sample boards. have changed the message to that.

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -143,6 +143,11 @@ private void login(boolean isBypassLogin) {
         getMainWindowHandle(mainStage.getTitle());
     }
 
+    private void openAndSaveSampleBoard(){
```
Is this the correct name for this method? The code is all about 'creating' the board, not opening or saving.

codemanmeet replied:
Yup, same here. it's `createAndLoadSampleBoard`.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,61 @@
+        type("test").push(KeyCode.TAB);
+        type("test");
+        click("Sign in");
+        clickMenu("File", "Logout");
```
Let me ask again: can lines 42-47 be written at a higher level? e.g. 
```
login();
logout();
```

codemanmeet replied:
I have abstracted it within `StartupBoardLauncherTest`.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,69 @@
+        type("test").push(KeyCode.TAB);
+        type("test");
+        click("Sign in");
+    }
```
Generalize it into `void login(owner, repo, password)` and pushed to parent class for reuse by others??

damithc replied:
BTW, name methods to sound like actions. e.g. `loginUsingDummyData`

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,69 @@
+    public void testSampleBoardOnLaunch(){
+        dummyLogin();
+        logout();
+
```
Why do we do a login and a logout? What happens after that?

codemanmeet replied:
We check the stored sample board and its corresponding panel details after that.

damithc replied:
Whoever reading the code needs to be guided through the code using comments. For starters, as a reader, I'm puzzled why a login and a logout is done at the beginning.

After that, I should see something like `assertEquals(expectedBoard, actualBoard)` but unfortunately we don't have the right asbtractions to do that, so we end up spraying assert statements all over the place.

codemanmeet replied:
Yup, as discussed before in the previous review, we need to incorporate `BoardInfo` class in our code. It requires quite a bit of refactoring for the existing board code. 

I have added comments now to explain the test scenario!

codemanmeet commented:
Prof., made all the relevant changes. Let me know if `mergeApproved` can be applied now!

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -150,6 +155,7 @@ private void disableUI(boolean disable) {
     }
 
     private void showMainWindow(String repoId) {
+        boolean isAFirstTimeUser = logic.getStoredRepos().isEmpty();
```
I thought I asked this line to be moved down? There is no point setting this variable until we need it, which happens at line 183.

codemanmeet replied:
I replied that this line cannot be moved down. `logic.openPrimaryRepository` changes `logic.getStoredRepos()` and hence the check has to be made exactly at this point. The boolean value obtained here is then used to decide whether the sample board should be opened.

damithc replied:
OK, that makes sense. No harm adding a comment to that effect.

codemanmeet replied:
Will do :+1:

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -105,14 +104,22 @@ public void sampleBoardAutoCreationTest() {
+        for (String panelName : BoardAutoCreator.getSamplePanelDetails().keySet()) {
+            assertEquals(panelInfos.get(i).getPanelName(), panelName);
+            assertEquals(panelInfos.get(i).getPanelFilter(), BoardAutoCreator.getSamplePanelDetails().get(panelName));
+            i++;
```
This method has high coupling because it takes more than it needs. Comparing boards doesn't need `PanelControl` or `BoardAutoCreator`. Can you do this instead? `verifySampeBoard(pc.getCurrentPanelInfos(), BoardAutoCreator.getSamplePanelDetails())`?

damithc commented on src/test/java/guitests/UITest.java:
```diff
@@ -361,6 +361,15 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
         }
     }
 
+    public void loginUsingDummyData(String owner, String repoName, String username, String password){
```
Why is it called `DaummyData`?

damithc replied:
This is a public method meant to be reused. Give a good header comment.

codemanmeet replied:
Okay

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,61 @@
+
+    private void logout(){
+        clickMenu("File", "Logout");
+    }
```
push to parent too?

codemanmeet replied:
Okay.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,61 @@
+    public void testSampleBoardOnLaunch(){
+        // First we login and logout for first time users.
+        loginUsingDummyData("dummy", "dummy", "test", "test");
+        logout();
```
Why logout?
How does the test ensure this is simulates a first time user?

codemanmeet replied:
It ensures that using `FXTestUtils.launchApp(TestUI.class, "--startupboard=true");`

codemanmeet replied:
We could do away with `logout` but I think it's always a good practice to tear down tests upon completion.

damithc replied:
If logout is part of teardown, then it should be put in the teardown method.

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -150,6 +155,8 @@ private void disableUI(boolean disable) {
 
     private void showMainWindow(String repoId) {
+        //Check whether there are existing stored repos. If not, it's indicative of first time users of Hubturbo.
+        boolean isAFirstTimeUser = logic.getStoredRepos().isEmpty();
```
The comment simply repeats what the statement says. What's missing is WHY this statement has to be top of the method. That was the confusion I had and that's what I'm trying to avoid in future readers.

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -105,14 +105,22 @@ public void sampleBoardAutoCreationTest() {
+    /**
+     * Tests whether the sample panels along with their filters are rightly created and present in the UI
+     */
+    public static void verifySampleBoard(PanelControl pc, Map<String, String> samplePanelDetails){
```
1. Why does the method take `PanelControl` as the parameter instead of `pc.getCurrentPanelInfos()`?
2. If you change it to take `pc.getCurrentPanelInfos()`, the method becomes more general, so the name/comment should not mention 'sample board'

Alternatively, you could feed both `PanelControl` and `BoardAutoCreator` into this method in which case all asserts should be done inside method. At the moment, some asserts are done outside and some inside.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,54 @@
+
+    @Test
+    public void testSampleBoardOnLaunch(){
+        //First we login and then logout for first time users.
```
This comment is redundant. Remember, we should comment about WHY, now HOW. The HOW should be obvious from the code.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,54 @@
+    public void testSampleBoardOnLaunch(){
+        //First we login and then logout for first time users.
+        login("dummy", "dummy", "test", "test");
+        logout();
```
Move the logout to teardown()?

damithc commented on src/test/java/guitests/UITest.java:
```diff
@@ -362,6 +362,29 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
+    }
+
+    /**
+     * Performs UI login on the dialog box upon startup.
```
Does it have to be `upon startup` or it can be anytime the login dialog is displayed?

damithc commented on src/main/java/ui/UI.java:
```diff
@@ -150,6 +155,8 @@ private void disableUI(boolean disable) {
     }
 
     private void showMainWindow(String repoId) {
+        //logic.openPrimaryRepository changes logic.getStoredRepos so the check for first time users is done here.
```
What do you mean `changes logic.getStoredRepos`?
```
//We infer this is the first time HT is being used if there are no repo data stored at the start up.
//This check needs to be done at the very beginning of the startup, before HT downloads any repo data.
```

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -105,14 +105,21 @@ public void sampleBoardAutoCreationTest() {
-            assertEquals(panelInfos.get(i).getPanelFilter(), SAMPLE_PANEL_FILTERS.get(i));
-            assertEquals(panelInfos.get(i).getPanelName(), SAMPLE_PANEL_NAMES.get(i));
+    /**
+     * Tests whether the panels along with their filters are rightly created and present in the UI
```
No it doesn't. The current implementation checks if the panels specified in one parameter is a subset of the other.

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -105,14 +105,21 @@ public void sampleBoardAutoCreationTest() {
         click("OK");
 
-        assertEquals(panelControl.getPanelCount(), SAMPLE_PANEL_NAMES.size());
+        assertEquals(panelControl.getPanelCount(), BoardAutoCreator.getSamplePanelDetails().size());
```
This can be done inside the `verifyBoard` method right?

codemanmeet replied:
Okay, will do

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,53 @@
+
+        assertEquals(testPref.getAllBoardNames().get(0), SAMPLE_BOARD);
+        verifyBoard(panelControl.getCurrentPanelInfos(), BoardAutoCreator.getSamplePanelDetails());
+    }
```
Doesn't this test make the one in `BoardAutoCreatorTest.java` redundant?

codemanmeet replied:
Both are needed as that one checks whether the board was created correctly through the `Auto Create` menu. This one is for new users.

damithc commented on src/test/java/guitests/UITest.java:
```diff
@@ -362,6 +362,29 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
+     * @param owner
+     * @param repoName
+     * @param username
+     * @param password
```
Side question: can you check if these 4 lines add any value when the IDE generates popup documentation for method calls?

codemanmeet replied:
Yup, it is useful. The @param is actually auto-generated by Intellij (which performs this according to Javadoc standards). The documentation appears like this - 
``java
guitests.UITest
public void login(String owner,
                  String repoName,
                  String username,
                  String password)
Performs UI login on the login dialog box.
Parameters:
owner - The owner of the repo.
repoName - The repository name
username - The Github/dummy username
password - The Github/dummy password
``

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -104,15 +104,23 @@ public void sampleBoardAutoCreationTest() {
-            assertEquals(panelInfos.get(i).getPanelName(), SAMPLE_PANEL_NAMES.get(i));
+    /**
+     * Tests panelControl properties associated with panels.
+     * panelDetails obtained from Boards are also tested as a subset of the panels in PanelControl.
```
Why a subset? I said 'subset' earlier because that was what the method was doing. Is this method checking for subset or exact match?

codemanmeet replied:
It is testing the exact match.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+    @Test
+    public void boardCreation_firstTimeUser_sampleBoardCreated(){
+        login("dummy", "dummy", "test", "test");
+        teardown();
```
Why call teardown here? teardown is supposed to happen at the end of the test. Instead of calling teardown, use a junit annotation for the teardown method.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertEquals(testPref.getAllBoardNames().size(), 1);
+        assertEquals(testPref.getAllBoardNames().get(0), SAMPLE_BOARD);
+        //Verifies the details of the sample board created.
+        verifyBoard(panelControl, BoardAutoCreator.getSamplePanelDetails());
```
Add blank lines between logical blocks e.g.
```java
login("dummy", "dummy", "test", "test");

//Ensures that only 1 board was created and it was the sample board
assertEquals(testPref.getAllBoardNames().size(), 1);
assertEquals(testPref.getAllBoardNames().get(0), SAMPLE_BOARD);
 
//Verifies the details of the sample board created.
verifyBoard(panelControl, BoardAutoCreator.getSamplePanelDetails());
```

damithc commented on src/test/java/guitests/UITest.java:
```diff
@@ -362,6 +362,29 @@ private void awaitCondition(Callable<Boolean> condition, int timeoutInSeconds) {
+    }
+
+    /**
+     * Performs UI login on the login dialog box. Assumes the cursor is on the login dialog box but nothing selected.
```
Is this condition really necessary ` box but nothing selected`?

codemanmeet replied:
Nope it's not necessary. If nothing is selected, `selectAll` will select the whole filter text anyways.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+public class StartupBoardLauncherTest extends UITest{
+
+    private PanelControl panelControl;
+    private Preferences testPref;
```
Any reason why this is called `testPref`?

codemanmeet replied:
It's a private preferences variable called in the test class, hence `testPref`. It's called the same in a number of other test classes like `RepositorySelectorTest`, `MenuControlTest` etc.

damithc replied:
By the same logic, the previous variable should be called `testPanelControl`? Is it a preference object specifically created for testing? If that is the case, it can be called `testPref`. If it is a regular preference object that happens to be used during testing, then I don't see why it is prefixed with `test` when other variables are not.

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+        teardown();
+        //Ensures that only 1 board was created and it was the sample board
+        assertEquals(testPref.getAllBoardNames().size(), 1);
+        assertEquals(testPref.getAllBoardNames().get(0), SAMPLE_BOARD);
```
See if there is a way to do this two things in one shot. I have a feeling there is a more elegant way to check the list of boards contains only sample board. I could be wrong.

damithc commented on src/test/java/guitests/BoardAutoCreatorTest.java:
```diff
@@ -104,15 +104,23 @@ public void sampleBoardAutoCreationTest() {
-            assertEquals(panelInfos.get(i).getPanelName(), SAMPLE_PANEL_NAMES.get(i));
+    /**
+     * Tests panelControl properties associated with panels.
+     * PanelDetails obtained from Boards are tested against current Panel infos as well.
```
I didn't think writing a proper comment for a method takes this many iterations :-p
```
Confirms the currently displayed board consists the set of panels specified in panelDetails
```
May need to rename the method to match the description `verifyCurrentBoard`

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+        UI ui = TestController.getUI();
+        panelControl = ui.getPanelControl();
+        testPref = UI.prefs;
+    }
```
looks like you don't need these three variables at all. 
Just use this at line 41?
```java
List<String> boardNames = TestController.getUI().getPanelControl().getAllBoardNames();
```
You can probably follow it up with this (not sure)
```java
assertEquals(boardNames, Arrays.asList(new String[]{SAMPLE_BOARD})
```

damithc commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,47 @@
+    @Test
+    public void boardCreation_firstTimeUser_sampleBoardCreated(){
+        login("dummy", "dummy", "test", "test");
+        // Workaround since we are unable to synchronize board creation on Travis post login through the dialog box.
```
I don't think anybody who didn't listen to your conversation with darius will be able to understand this phrase `unable to synchronize board creation on Travis post login through the dialog box` 
Comments are not 'notes to self'. Someone else should be able to make sense out of them.
@dariusf why can't logout be at line 46 instead of here?

dariusf replied:
I'm not entirely sure, but I'm working with Manmeet on the issue. Our conversation earlier was on a slightly different problem.

codemanmeet replied:
After discussing with @dariusf , it's been rectified through `countPanelsShown` which takes a `FutureTask<Integer>` and performs a `PlatformEx.runAndWait` on it. `TestUtils.awaitCondition` is then run on it until all the panels are shown on the UI. Thanks @dariusf !

dariusf commented on src/test/java/guitests/StartupBoardLauncherTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    public void boardCreation_firstTimeUser_sampleBoardCreated(){
+        login("dummy", "dummy", "test", "test");
+        TestUtils.awaitCondition(() -> BoardAutoCreator.getSamplePanelDetails().size() == countPanelsShown());
```
@damithc we've removed the `logout()` by synchronising properly when checking UI state. I'm hoping to add an abstraction for this similar in spirit to [this](https://github.com/HubTurbo/HubTurbo/blob/952-nondeterministic-tests/src/test/java/guitests/UITest.java#L360-L366), but haven't worked out details and best practices yet. Could we leave this here for now (with a comment?) and abstract it away when we know what exactly is required?

damithc replied:
Sure. This line shows we are waiting for something to complete, which is not surprising in UI tests and is unlikely to confuse the reader. An unexpected logout with an incomprehensible comment is not.

dariusf commented:
@codemanmeet this is ready. Could you squash the unnecessary commits first? Also coverage isn't working for some reason.
